### DISCUSSION TITLE
allow modern enum class/scoped enums with type specifier and assigning values

### DISCUSSION THREAD

**Discussion by JSG-0** [eProsima/Fast-DDS-Gen#293] [Ideas]
Scoped enums have been available since C++-11 (13 years ago). Other DDS implementations support this in their IDL.
The advantages are well documented.

Add IDL support for scoped enums a.k.a. "enum classes" with a type specified "enum base" that can be any cstdint and allow specifying a value for each enumeration independently, including negative values, and since scoped allow different enums to reuse label names.

**Comment by JLBuenoLopez** [comment]
@JSG-0

Thanks for your contribution. According to [Fast DDS CONTRIBUTING guidelines](https://github.com/eProsima/Fast-DDS/blob/master/CONTRIBUTING.md#issues-and-support) I am moving this ticket to the Ideas forum where feature requests are discussed.

On the one hand, enum class support has been introduced in #271 into the development branch [feature/xtypes_typeobject](https://github.com/eProsima/Fast-DDS-Gen/tree/feature/xtypes_typeobject). This branch will be merged and released soon.

On the other hand, I am afraid that support for `@bit_bound` builtin annotation that allows modifying the enum base and the `@value` builtin annotation that allows setting the value of each enumeration literal, is not in the current roadmap. You might contact [eProsima's commercial support team](https://github.com/eProsima/Fast-DDS#getting-help) to discuss options to accelerate their development.

Regarding the scope, this is something already supported in the current Fast DDS-Gen. Nevertheless I have issued eProsima/dds-types-test#22 to add this specific test case into Fast DDS-Gen CI. The CI is updated in #292.

**Reply by JSG-0** [reply]
Thanks for the quick response: dealing with large volume of real-time messages with lots of enums, and never need more than 8 bits. Being forced into uint32 is a big waste. And there are interface requirements specifying negative enum values and with very specific non-monitorically increasing values.

**Reply by JSG-0** [reply]
Is it on the roadmap yet to allow me to specify custom, random values for enums yet? Including negative values?

**Reply by richiware** [reply]
Added support in #379
